community/index/src/main/java/org/neo4j/index/gbptree/InternalTreeLogic.java

```diff
@@ -486,16 +528,41 @@ public VALUE remove( PageCursor cursor, KEY key, VALUE into, long stableGenerati
         }
 
         long currentId = cursor.getCurrentPageId();
-        goTo( cursor, bTreeNode.childAt( cursor, pos, stableGeneration, unstableGeneration ) );
+        long childId = bTreeNode.childAt( cursor, pos, stableGeneration, unstableGeneration );
+        goTo( cursor, childId, stableGeneration, unstableGeneration );
 
-        VALUE result = remove( cursor, key, into, stableGeneration, unstableGeneration );
+        VALUE result = remove( cursor, structurePropagation, key, into, stableGeneration, unstableGeneration );
 
-        goTo( cursor, currentId );
+        goTo( cursor, currentId, stableGeneration, unstableGeneration );
+        if ( structurePropagation.hasNewGen )
+        {
+            structurePropagation.hasNewGen = false;
+            bTreeNode.setChildAt( cursor, structurePropagation.left, pos, stableGeneration, unstableGeneration );
+        }
 
         return result;
     }
 
-    private VALUE removeFromLeaf( PageCursor cursor, KEY key, VALUE into )
+    /**
+     * Remove given {@code key} and associated value from tree if it exists. The removed value will be stored in
+     * provided {@code into} which will be returned for convenience.
+     * <p>
+     * If the given {@code key} does not exist in tree, return {@code null}.
+     * <p>
+     * Leaves cursor at same page as when called. No guarantees on offset.
+     *
+     * @param cursor {@link PageCursor} pinned to page where remove is to be done.
+     * @param structurePropagation {@link StructurePropagation} used to report structure changes between tree levels.
+     * @param key key to be removed
+     * @param into {@code VALUE} instance to write removed value to
+     * @param stableGeneration stable generation, i.e. generations <= this generation are considered stable.
+     * @param unstableGeneration unstable generation, i.e. generation which is under development right now.
+     * @return Provided {@code into}, populated with removed value for convenience if {@code key} was removed.
+     * Otherwise {@code null}.
+     * @throws IOException on cursor failure
+     */
+    private VALUE removeFromLeaf( PageCursor cursor, StructurePropagation<KEY> structurePropagation,
+            KEY key, VALUE into, long stableGeneration, long unstableGeneration ) throws IOException
```

Comment:
Could `StructurePropagation` be used for the out-parameter here, or is that conflating concerns? Value types sounds like they could be of some use here, if we had them.